test/method.jl

```diff
@@ -0,0 +1,74 @@
+using Test
+
+@testset "Normal generated functions" begin
+    @generated g1(x) = :(x)
+    @test g1(1) == 1
+    @test g1("hi") == "hi"
+end
+
+# Invalidating Generated Functions
+# TODO: For some reason this doesn't work inside a testset right now (See below)
+@generated foo() = bar()
+bar() = 2
+# We can still call foo() even though bar() was defined after! Hooray! :)
+@test foo() == 2
+# Now we can change bar(), and foo() is also updated! Woohoo! :D
+bar() = 3
+@test foo() == 3
+
+
+@testset "invalidating generated functions in a testset" begin
+    @generated foo() = bar()
+    bar() = 2
+
+    # TODO: It seems like this doesn't work becuase of the @testset. Is that expected?
+    # Would this work for regular functions? I think it's broken...
+    @test foo() == 2
+    bar() = 3
+    @test_broken foo() == 3
+end
+
+
+# Functions that take arguments
+@generated f(x) = f2(x) + f2(x)
+f2(x::Type) = sizeof(x)
+@test f(1) == 16
+f2(x::Type) = sizeof(x)÷2
+@test f(1) == 8
+
+
+# Method at bottom of call-stack accepts ::Type, not ::Any
+# The simple case -- bar(::Any):
+@generated foo(x) = bar(x)
+bar(x) = 2
+@test foo(1) == 2
+bar(x) = 3
+@test foo(1) == 3
+# This also works, with t(::Type{Int})
+@generated f_type(x) = t(x)
+t(::Type{Int}) = 2
+@test f_type(1) == 2
+t(::Type{Int}) = 3
+@test f_type(1) == 3
+# Yet for some reason this does not work:
+# Somehow having t1(T) call typemax prevents forming a backedge from t1 to the generator.
+@generated f_type2(x) = t1(x)
+t1(T) = typemax(T)
+@test f_type2(1) == typemax(Int)
+t1(T) = 3
+@test_broken f_type2(1) == 3
```

Comment:
This is one of the currently broken examples. Somehow, when the generator is compiled, it doesn't cause any backedges to be set from `t1` to it.

This simple utility function prints out all the backedges for all specializations of a function, and you can see that there are none for `t1`:
```julia
julia> NHDalyUtils.func_all_backedges(t1)
1-element Array{Pair{Any,Any},1}:
 :MethodTable => Any[]
```
Whereas the previous, working example _does_ get such backedges set:
```julia
julia> NHDalyUtils.func_all_backedges(t)
2-element Array{Pair{Any,Array{Any,1}},1}:
                 :MethodTable => [Tuple{typeof(t),Any}, MethodInstance for #s5#4(::Any, ::Any)]
 Tuple{typeof(t),Type{Int64}} => [MethodInstance for #s5#4(::Any, ::Any)]
```

Reply:
After some digging, I think I've found a MWE that explains why this case doesn't work, but it boggles my understanding of how julia recompiles functions...
As I show in this example, functions that call `typemax` don't get backedges set from typemax, despite the fact that they _do_ inline the result.
But by some mysterious magic, they still get recompiled when `typemax` is updated, so somehow there's something other than backedges that can cause function recompilation??

```julia
julia> struct X x end

julia> Base.typemax(::Type{X}) = 8

julia> f1(t) = Base.typemax(t)
f1 (generic function with 1 method)

julia> f1(X)
8

julia> Base.method_instances(typemax, (Type{X},))[1].backedges
ERROR: UndefRefError: access to undefined reference
Stacktrace:
 [1] getproperty(::Any, ::Symbol) at ./sysimg.jl:18

julia> f2(x::T) where T = Base.typemax(T)  # This function _does_ get a backedge added to typemax.
f2 (generic function with 1 method)

julia> f2(X(2))
8

julia> Base.method_instances(typemax, (Type{X},))[1].backedges
1-element Array{Any,1}:
 MethodInstance for f2(::X)

julia> Base.typemax(::Type{X}) = 100

julia> f1(X)  # wat! if there is no backedge to f1 from typemax, how is this getting recompiled?!?
100

julia> @code_typed f1(X)  # But we know it _is_ getting recompiled, because the value is inlined, here:
CodeInfo(
1 ─     return 100
) => Int64

julia> methods(typemax).mt.backedges  # And it's not hiding on the MethodTable, as we can see.
ERROR: UndefRefError: access to undefined reference
```

So my test is failing because there's no backedge from `typemax` to `f1`. And in my test case above, `f1` is the generator body. So the generator body is never properly invalidated, and so my generated function is never re-generated.

But somehow typemax _is_ able to cause `f1` to recompile via some mystery voodoo that i don't understand. And so I guess I want to also apply that to my generator body?


-------

@vtjnash you pointed to this test case as an example of why "this approach won't work". Is this what you were referring to? That sometimes there is a different mechanism besides backedges that is used to trigger function recompilation?

Reply:
This is now explained via my explanation below, https://github.com/JuliaLang/julia/pull/32774#issuecomment-520471641.

It turns out that -- despite what it looks like -- the call to `typemax(t)` in `f1` above is triggering a _dynamic dispatch_, so as explained in that comment, the backedges mechanism implemented in this PR so far has no way to trigger regenerating the generated function. (It _looks like_ the call to typemax is being inlined above, but this is actually due to `@code_typed` showing fully specialized code, even though it isn't actually fully specialized in reality: https://github.com/JuliaLang/julia/issues/32834.)

A simpler example that showcases why this doesn't work can be found here -- this example fails for the current implementation in this PR (as of f20d374):
```julia
baz() = 2
f() = Any[baz][1]()  # f calls `baz()` through a type-erased `Any` reference, and inference cannot de-virtualize it.
@staged foo() = f()
@test foo() == 2  # When computing the result for `foo()`, `baz()` is called via dynamic dispatch
baz() = 4  # Updating baz does not trigger regenerating foo
@test foo() == 4  # TEST FAILURE: foo() still returns 2
```